features/src/main/scala/com/salesforce/op/features/types/FeatureTypeSparkConverter.scala

```diff
@@ -147,14 +147,36 @@ case object FeatureTypeSparkConverter {
 
       // Numerics
       case wt if wt <:< weakTypeOf[t.RealNN] => (value: Any) =>
-        if (value == null) None else Some(value.asInstanceOf[Double])
+        value match {
+          case null => None
+          case _: Float => Some(value.asInstanceOf[Float].toDouble)
+          case _: Double => Some(value.asInstanceOf[Double])
+          case _ => throw new IllegalArgumentException(s"RealNN type mapping is not defined for ${value.getClass}")
+        }
       case wt if wt <:< weakTypeOf[t.Real] => (value: Any) =>
-        if (value == null) FeatureTypeDefaults.Real.value else Some(value.asInstanceOf[Double])
+        value match {
+          case null => FeatureTypeDefaults.Real.value
+          case _: Float => Some(value.asInstanceOf[Float].toDouble)
+          case _: Double => Some(value.asInstanceOf[Double])
+          case _ => throw new IllegalArgumentException(s"Real type mapping is not defined for ${value.getClass}")
+        }
       case wt if wt <:< weakTypeOf[t.Integral] => (value: Any) =>
-        if (value == null) FeatureTypeDefaults.Integral.value else Some(value.asInstanceOf[Long])
+        value match {
+          case null => FeatureTypeDefaults.Integral.value
+          case _: Short => Some(value.asInstanceOf[Short].toLong)
+          case _: Int => Some(value.asInstanceOf[Int].toLong)
+          case _: Long => Some(value.asInstanceOf[Long])
+          case _ => throw new IllegalArgumentException(s"Integral type mapping is not defined for ${value.getClass}")
+        }
       case wt if wt <:< weakTypeOf[t.Binary] => (value: Any) =>
         if (value == null) FeatureTypeDefaults.Binary.value else Some(value.asInstanceOf[Boolean])
 
+      // Date & Time
```

Comment:
Why do you need special handling for `Date` and `DateTime`? were you planning to take care of `TimestampType` spark type?

Reply:
Since Date and DateTime types in TransmogrifAI inherit from Integral types, their values are captured by this case `case wt if wt <:< weakTypeOf[t.Integral]` and it works, because `<:<` operator on type tags check inheritance.

Reply:
If you would like to mape spark types `DateType` and `TimestampType` into TransmogrifAI `Date` and `DateTime` types accordingly then you should move the cases `case wt if wt <:< weakTypeOf[t.Date]` and `case wt if wt <:< weakTypeOf[t.DateTime]` before the numeric checks.